src/txmempool.cpp

```diff
@@ -764,48 +918,174 @@ void CTxMemPool::check(const CCoinsViewCache& active_coins_tip, int64_t spendhei
+                fee += it->get().GetModifiedFee();
+                size += it->get().GetTxSize();
+                // Check that all parents are in txs_so_far
+                for (const auto& parent : it->get().GetMemPoolParentsConst()) {
+                    assert(txs_so_far.count(parent));
+                }
+                txs_so_far.insert(*it);
+            }
+            assert(fee == cluster->m_chunks[i].fee);
+            assert(size == cluster->m_chunks[i].size);
+        }
+        innerUsage += cluster->GetMemoryUsage();
+    }
+
+    // Check that each cluster is connected.
+    for (const auto & [id, cluster] : m_cluster_map) {
+        // Since we've checked the parents and children already, we'll use
+        // those values here.
+        // We'll check that if we walk to every transaction reachable from the
+        // first one, that we get every tx in the cluster.
+        auto first_tx = cluster->m_chunks.front().txs.front();
+        std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef> work_queue;
+        int reachable_txs = 1; // we'll count the transactions we reach.
+
+        WITH_FRESH_EPOCH(m_epoch);
+        visited(first_tx.get());
+        assert(first_tx.get().GetMemPoolParentsConst().size() == 0); // first tx can never have parents.
+        for (auto child : first_tx.get().GetMemPoolChildrenConst()) {
+            work_queue.push_back(child);
+            visited(child.get());
+            ++reachable_txs;
+        }
+        while (work_queue.size() > 0) {
+            auto next_tx = work_queue.back();
+            work_queue.pop_back();
+            for (auto parent : next_tx.get().GetMemPoolParentsConst()) {
+                if (!visited(parent.get())) {
+                    ++reachable_txs;
+                    work_queue.push_back(parent);
+                }
+            }
+            for (auto child : next_tx.get().GetMemPoolChildrenConst()) {
+                if (!visited(child.get())) {
+                    ++reachable_txs;
+                    work_queue.push_back(child);
+                }
+            }
+        }
+        assert(reachable_txs == cluster->m_tx_count);
+    }
+
     assert(totalTxSize == checkTotal);
     assert(m_total_fee == check_total_fee);
     assert(innerUsage == cachedInnerUsage);
 }
 
-bool CTxMemPool::CompareDepthAndScore(const uint256& hasha, const uint256& hashb, bool wtxid)
+// Return true if a comes before b in mempool sort order
+bool CTxMemPool::CompareMiningScore(txiter a, txiter b) const
+{
+    if (a == b) return false; // An element cannot be less than itself.
+
+    CAmount a_fee = a->m_cluster->m_chunks[a->m_loc.first].fee;
+    int64_t a_size = a->m_cluster->m_chunks[a->m_loc.first].size;
+    CAmount b_fee = b->m_cluster->m_chunks[b->m_loc.first].fee;
+    int64_t b_size = b->m_cluster->m_chunks[b->m_loc.first].size;
+
+    int64_t a_score = a_fee * b_size;
+    int64_t b_score = b_fee * a_size;
```

Comment:
It would be good to explain the rationale for the miner score somewhere. Why a * b and b * a?